_episodes/l3_03_test_fixture.md

```diff
@@ -0,0 +1,505 @@
+
+```cpp
+class Your_test_fixture_class_name : public::testing::Test {
+    public:
+        ClassUnderTest publicInstance;
+
+    protected:
+        ClassUnderTest protectedInstance;
+
+    private:
+        ClassUnderTest privateInstance;
+};
+```
+
+Please note that you do not need to use all three access specifiers `public`, `private` and `protected`  defined above. The choice would depend on the following:-
+
+1. **public**: This is the most commonly used access specifier in test fixtures. It allows the test fixture class and its members (including the instance of the class we want to test) to be accessed from anywhere, including test cases defined outside the fixture class.
+
+2. **protected**: This access specifier restricts the visibility of the test fixture class and its members to derived classes and other classes within the same hierarchy. It can be useful if we have derived test fixture classes that need access to the class under test or if we want to limit the accessibility of the test fixture within a certain scope.
+
+3. **private**: This access specifier restricts the visibility of the test fixture class and its members only to the test fixture class itself. It can be useful if we want to encapsulate the test logic within the test fixture class and prevent external access or if we want to limit the scope of the test fixture.
+
+For our course, we will be using the `public` access specifier. Once we have the test fixture class, we use the `TEST_F` macro available in GoogleTest to write our tests instead of the `TEST` macro we have been using so far. The general syntax is given below.
+
+```cpp
+TEST_F(Your_test_fixture_class_name, Your_test_name) {
+    // Test logic goes here
+}
+```
+
+Since we now have all the basic tools to create our own test fixtures, let us rewrite the above tests by using a fixture. The code is present in this [2_employeetest.cpp](../code/Chapter3/2_employeetest.cpp). For reference, the tests are shown in the cell below.
+
+```cpp
+// Create a test fixture.
+class EmployeeTestFixture : public::testing::Test {
+    public:
+        Employee employee{"John", 25, 45000, 12, 5000};
+
+};
+
+// Test if we can set the name of an employee.
+TEST_F(EmployeeTestFixture, CanSetName) {
+    employee.setName("John Doe");
+    EXPECT_EQ(employee.getName(), "John Doe");
+}
+
+// Test that the name cannot be empty.
+TEST_F(EmployeeTestFixture, NameCannotBeEmpty) {
+    EXPECT_THROW(employee.setName(""), invalid_argument);
+}
+
+// Test if we can set the age of an employee.
+TEST_F(EmployeeTestFixture, CanSetAge) {
+    employee.setAge(30);
+    EXPECT_EQ(employee.getAge(), 30);
+}
+
+//Test that the tax calculation is correct.
+TEST_F(EmployeeTestFixture, TaxCalculationIsCorrect) {
+    EXPECT_EQ(employee.getTaxAmount(), 7500);
+}
+
+// Check that the net salary is correct.
+TEST_F(EmployeeTestFixture, NetSalaryIsCorrect) {
+    EXPECT_EQ(employee.getNetSalary(), 43500);
+}
+```
```

Comment:
Maybe just have the `CanSetName` and `CanSetAge` examples, like for the non-fixture example above. Demonstrates the point clearly and concisely.

Reply:
Implemented this suggestion and removed these two functions in 8d4ec5ce637e86c5d3f9ded79606867d056496a5